### ISSUE TITLE
the yield generator should have by default a random fee

### ISSUE THREAD

**Issue by greenaddress** [JoinMarket-Org/joinmarket#632]
To make it harder to track things on the network it seems to make sense to use a random fee (within a range/max value?)

**Comment by AlexCato** [comment]
Can easily be done, though arguably, as long as there's others not changing the default, individual yieldgens are still hard to be distinguished by the fee alone. But it might add one more layer, so it's possibly a slight improvement.
Could offer a PR to randomize the default relative cj fee to be between 0.018% and 0.022% (currently: fixed 0.02% ) or something along those lines?

**Comment by AdamISZ** [comment]
Yeah except this is to be added to a reduction in the default imo. Obv people like to set their own prices (that's the point) but having a *default* tweak I thought, on balance after discussion with @greenaddress was a sound idea. People are lazy and run with the same fee all the time, mostly. Well, at least with the default scripts.

**Comment by juscamarena** [comment]
Wouldn't it be better to have a random fee set in a specific range after every cj or every x cjs?

**Comment by AlexCato** [comment]
That would complicate things, as you'd have to announce changed orders in public whenever the cj fee changes.
The current yield generator's goal is to announce his offerings in public as rarely as possible to avoid some information leakage about coinjoins (thats why it touches the maximum mixdepth only if absolutely necessary, because this leads to a changed maximum offer size). Could be changed only if an new announcement had to be made anyways to avoid this problem.

But I dont see what's gained by changing this as well: a snoop would monitor the channel 24/7 anyways, and random cj fees dont help at all with that, as your bot's nickname stays the same and offers can therefor easily be linked just by that.

**Comment by AdamISZ** [comment]
I thought about this more carefully. TLDR I think we should strongly enforce a different fee value each time (of course "enforce" is not possible, but you get it).

First let's be clear that there's 2 different threat models; one is real time channel monitoring (which admittedly is easy to do), and the other is blockchain-only analysis. 

Now, consider the latter - suppose after the event, blockchain-only analysis. Considering one tx only, the size of the fee doesn't really matter, since the inputs and change outputs can (almost) always be linked using subset-sum analysis, while the coinjoin outs cannot, no matter what the fee is. But a single entity always using a very unique fee, let's say 0.0618%, gives a watermark to his own (inputs, change) set for each tx he takes part in, assuming that within some tolerance no one else is near to that value, which will effectively link his transactions that send to different mixdepths (by our design this is what we're trying to hide).

A lot of people are naturally choosing simple round numbers, and sometimes defaults, which is good of course, but from this (blockchain analysis) point of view I think the "enforce simple round numbers" model is inferior to the "enforce changing values each time". It's actually rather tricky to find the optimal way to do this without impacting users' ability to make a market choice too much; perhaps some kludge of the two.

Threat 1, real time monitoring: for that then I agree it's much better if we have fixed, unchanging fee values for one bot (avoid reannounce), but that would leave us needing to enforce round numbers with little choice. I can't now see for sure what kind of tradeoff is right, or whether there is a better, non-tradeoff way...

**Comment by juscamarena** [comment]
Exactly. That's the reasoning I had when I wrote that, but excluded adding that bit. I don't really know all the working parts yet, but I understand the blockchain implications much more atm.

**Comment by AdamISZ** [comment]
This interacts with another discussion we had recently.

Consider the complex bots like -oscillator and -deluxe. These would re-announce with complex algorithms and various times (not just at tx time). I remember saying on IRC I could see value in both extremes: either heavily avoid reannounce (this is the yg-pe.py strategy) or randomly reannoucing at non-tx times, with varying fees and amounts.

The discussion above, I think, strongly favours the former: use complex un-disentanglable reannouncement strategies with fees varying at least enough such that each transaction has a different fee. The only reason I was reluctant about this before (apart from complexity) was the potentially heavy additional messaging load, but I suspect it's not that bad (the biggest load is actually in !tx and maybe orderbook request response).

So now I'm strongly leaning in that direction.
